File: Sources/NBT/ConsecutiveNBTCodec.swift
import Foundation

struct ConsecutiveNBTRecord {
    var document: NBTDocument
    var rawData: Data
    var encoding: NBTEncoding
}

enum ConsecutiveNBTCodec {
    static func decode(_ data: Data) throws -> [ConsecutiveNBTRecord] {
        guard !data.isEmpty else { return [] }

        var cursor = BinaryCursor(data: data)
        var records = [ConsecutiveNBTRecord]()
        while cursor.remaining > 0 {
            let remaining = data[cursor.offset..<data.count]
            if remaining.allSatisfy({ $0 == 0 }) { break }

            let before = cursor.offset
            do {
                let document = try BedrockNBTCodec.decodeDocument(cursor: &cursor, encoding: .littleEndian)
                records.append(ConsecutiveNBTRecord(
                    document: document,
                    rawData: data.subdata(in: before..<cursor.offset),
                    encoding: .littleEndian
                ))
            } catch {
                if records.isEmpty {
                    var fallbackCursor = BinaryCursor(data: data)
                    do {
                        let fallback = try BedrockNBTCodec.decodeDocument(
                            cursor: &fallbackCursor, encoding: .littleEndianVarInt)
                        let trailing = data[fallbackCursor.offset..<data.count]
                        guard trailing.allSatisfy({ $0 == 0 }) else {
                            throw MCBEEditorError.malformedData("VarInt NBT 后仍有未解析数据")
                        }
                        return [ConsecutiveNBTRecord(
                            document: fallback, rawData: data, encoding: .littleEndianVarInt)]
                    } catch {
                        throw MCBEEditorError.malformedData(
                            "NBT 既不是有效 Little Endian，也不是完整 VarInt：\(error.localizedDescription)")
                    }
                }
                throw MCBEEditorError.malformedData("连续 NBT 在偏移 \(before) 解析失败：\(error.localizedDescription)")
            }

            guard cursor.offset > before else {
                throw MCBEEditorError.malformedData("NBT 解析器没有前进")
            }
        }
        return records
    }

    static func encode(_ records: [ConsecutiveNBTRecord]) throws -> Data {
        var data = Data()
        for record in records {
            data.append(try BedrockNBTCodec.encode(record.document, encoding: record.encoding))
        }
        return data
    }
}

File: Windows/MCBEEditor.Core/Nbt/ConsecutiveNbtCodec.cs
namespace MCBEEditor.Core.Nbt;

public sealed record ConsecutiveNbtRecord(NbtDocument Document, byte[] RawData, NbtEncoding Encoding);

public static class ConsecutiveNbtCodec
{
    public static IReadOnlyList<ConsecutiveNbtRecord> Decode(ReadOnlySpan<byte> data)
    {
        if (data.Length == 0) return Array.Empty<ConsecutiveNbtRecord>();

        var records = new List<ConsecutiveNbtRecord>();
        var offset = 0;
        while (offset < data.Length)
        {
            var remaining = data[offset..];
            if (AllZero(remaining)) break;

            try
            {
                var document = BedrockNbtCodec.DecodeOne(remaining, out var consumed, NbtEncoding.LittleEndian);
                if (consumed <= 0) throw new InvalidDataException("NBT parser did not advance.");
                records.Add(new ConsecutiveNbtRecord(document, remaining[..consumed].ToArray(), NbtEncoding.LittleEndian));
                offset += consumed;
            }
            catch (Exception littleEndianError)
            {
                if (records.Count != 0)
                    throw new InvalidDataException($"Consecutive NBT failed at offset {offset}: {littleEndianError.Message}", littleEndianError);

                try
                {
                    var fallback = BedrockNbtCodec.Decode(remaining, NbtEncoding.LittleEndianVarInt);
                    var reencoded = BedrockNbtCodec.Encode(fallback, NbtEncoding.LittleEndianVarInt);
                    if (reencoded.Length > remaining.Length || !remaining[..reencoded.Length].SequenceEqual(reencoded) || !AllZero(remaining[reencoded.Length..]))
                        throw new InvalidDataException(
                            $"Little-endian VarInt fallback did not consume a canonical NBT payload at offset {offset}.",
                            littleEndianError);
                    return [new ConsecutiveNbtRecord(fallback, remaining.ToArray(), NbtEncoding.LittleEndianVarInt)];
                }
                catch (Exception)
                {
                    throw new InvalidDataException($"Consecutive NBT failed at offset {offset}: {littleEndianError.Message}", littleEndianError);
                }
            }
        }
        return records;
    }

    public static byte[] Encode(IEnumerable<ConsecutiveNbtRecord> records)
    {
        ArgumentNullException.ThrowIfNull(records);
        using var stream = new MemoryStream();
        foreach (var record in records)
        {
            var encoded = BedrockNbtCodec.Encode(record.Document, record.Encoding);
            stream.Write(encoded);
        }
        return stream.ToArray();
    }

    private static bool AllZero(ReadOnlySpan<byte> data)
    {
        foreach (var value in data) if (value != 0) return false;
        return true;
    }
}
